docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+1. As a cluster operator, I want my cluster architecture to be always consistent with best practice, in order to have reliable cluster provisioning without having to understand the details of underlying datastores, replication etc…
+1. As a cluster operator, I want to know if my cluster’s control plane is healthy in order to understand if I am meeting my SLOs with my end users.
+1. As a cluster operator, I want to be able to quickly respond to a Kubernetes CVE by upgrading my clusters in an automated fashion.
+1. As a cluster operator, I want to be able to quickly respond to a non-Kubernetes CVE that affects my base image or Kubernetes dependencies by upgrading my clusters in an automated fashion.
+1. As a cluster operator, I would like to upgrade to a new minor version of Kubernetes so that my cluster remains supported.
+1. As a cluster operator, I want to know why my cluster isn’t working properly after creation. I have ended up with an API server I can access, but CNI isn’t working, and new machines are not registering themselves with the control plane.
+
+#### Identified features from user stories
+
+1. Based on the function of kubeadm, the control plane provider must be able to scale the number of replicas of a control plane from 1 to X, meeting user stories 1 through 4.
+1. To address user story 5, the control plane provider should provide validation of the number of replicas in a control plane. Where the stacked etcd topology is used (i.e., in the default implementation), the number of replicas should be an odd number, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members). When external etcd is used, any number is valid.
+1. In service of user story 5, the control plane provider should also manage etcd membership via kubeadm as part of scaling up and down.
+1. The control plane provider should provide indicators of health to meet user story 6 and 10. This should include at least the state of etcd and information about which replicas are currently healthy or not. For the default implementation, health attributes based on artifacts kubeadm installs on the cluster may also be of interest to cluster operators.
+1. The control plane provider should be able to upgrade a control plane’s version of Kubernetes as well as updating the underlying machine image on cloud installations.
+
+### Implementation Details/Notes/Constraints
+
+#### New API Types
+
+ControlPlane:
+
+```go
+package v1alpha3
+
+import (
+    corev1  "k8s.io/api/core/v1"
+    metav1  "k8s.io/apimachinery/pkg/apis/meta/v1"
+    cabpkv1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
+)
+
+// ControlPlaneSpec defines the desired state of ControlPlane
+type ControlPlaneSpec struct {
+    // Number of desired machines. Defaults to 1.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    Replicas int32 `json:"replicas"`
+
+    // Version defines the desired Kubernetes version.
```

Comment:
Sure, but seems worthwhile to add it here so it can closely reflect what the implementation PR will look like, more or less?

Reply:
I think this should be out of scope for now, since we don't currently enforce any type of version validation, and anything we introduce could potentially affect the adoption of existing clusters as part of the upgrade.

This is probably something that should be a separate issue, since it affects the current uses of the exisiting version fields we have today.

Reply:
Opened https://github.com/kubernetes-sigs/cluster-api/issues/1629